File: go/sms.go
package msgine

import (
	"context"
	"net/url"
	"strings"
)

// SMSService provides methods for sending and listing SMS messages.
type SMSService struct {
	t *transport
}

// SendSmsRequest is the payload for sending an SMS.
// To accepts a comma-separated string or use SendSmsOptions.ToList for a slice.
type SendSmsRequest struct {
	To             string `json:"to"`
	Message        string `json:"message"`
	From           string `json:"from,omitempty"`
	ScheduledFor   string `json:"scheduledFor,omitempty"`
	IdempotencyKey string `json:"idempotencyKey,omitempty"`
}

// Send sends one or more SMS messages. The To field should be a comma-separated
// list of E.164 phone numbers (e.g. "+256700000000,+256712345678").
func (s *SMSService) Send(ctx context.Context, opts SendSmsOptions) (*SmsMessage, error) {
	if len(opts.To) == 0 {
		return nil, &ValidationError{Field: "to", Message: "at least one recipient is required"}
	}
	if opts.Message == "" {
		return nil, &ValidationError{Field: "message", Message: "message is required"}
	}
	if len(opts.Message) > 1000 {
		return nil, &ValidationError{Field: "message", Message: "message must be 1000 characters or fewer"}
	}
	if opts.From != "" && len(opts.From) > 11 {
		return nil, &ValidationError{Field: "from", Message: "sender ID must be 11 characters or fewer"}
	}

	body := SendSmsRequest{
		To:             strings.Join(opts.To, ","),
		Message:        opts.Message,
		From:           opts.From,
		ScheduledFor:   opts.ScheduledFor,
		IdempotencyKey: opts.IdempotencyKey,
	}
	var result SmsMessage
	if err := s.t.do(ctx, "POST", "/developers/sms", nil, body, &result); err != nil {
		return nil, err
	}
	return &result, nil
}

// History returns a paginated list of sent SMS messages.
func (s *SMSService) History(ctx context.Context, page, limit int) (*PaginatedResult[SmsMessage], error) {
	q := queryValues("page", page, "limit", limit)
	var result PaginatedResult[SmsMessage]
	if err := s.t.do(ctx, "GET", "/developers/sms/history", q, nil, &result); err != nil {
		return nil, err
	}
	return &result, nil
}

// Get returns a single SMS message by ID.
func (s *SMSService) Get(ctx context.Context, id string) (*SmsMessage, error) {
	if id == "" {
		return nil, &ValidationError{Field: "id", Message: "id is required"}
	}
	var result SmsMessage
	if err := s.t.do(ctx, "GET", "/developers/"+url.PathEscape(id), nil, nil, &result); err != nil {
		return nil, err
	}
	return &result, nil
}

File: src/modules/sms.ts
import { HttpClient, toRecipientString } from '../http-client';
import {
  HttpMethod,
  MsGineValidationError,
  PaginationOptions,
  SendSmsOptions,
  SmsHistoryResult,
  SmsMessage,
} from '../types';

export class SmsModule {
  constructor(private readonly http: HttpClient) {}

  /**
   * Send an SMS to one or more recipients.
   *
   * @example
   * ```ts
   * const msg = await client.sms.send({
   *   to: ['+256701234567', '+256702345678'],
   *   message: 'Your OTP is 123456',
   * });
   * console.log(msg.id, msg.status);
   * ```
   */
  async send(options: SendSmsOptions): Promise<SmsMessage> {
    validate(options);

    return this.http.request<SmsMessage>({
      method: HttpMethod.POST,
      path: '/developers/sms',
      body: {
        to: toRecipientString(options.to),
        message: options.message,
        from: options.from,
        scheduledFor: options.scheduledFor,
        idempotencyKey: options.idempotencyKey,
      },
    });
  }

  /**
   * Get paginated SMS history.
   *
   * @example
   * ```ts
   * const { items, total } = await client.sms.history({ page: 1, limit: 20 });
   * ```
   */
  async history(options: PaginationOptions = {}): Promise<SmsHistoryResult> {
    const raw = await this.http.request<{
      messages: SmsMessage[];
      total: number;
      page: number;
      pages: number;
    }>({
      method: HttpMethod.GET,
      path: '/developers/sms/history',
      queryParams: { page: options.page ?? 1, limit: options.limit ?? 20 },
    });

    return { items: raw.messages, total: raw.total, page: raw.page, pages: raw.pages };
  }
}

// ── Validation ────────────────────────────────────────────────────────────────

function validate(o: SendSmsOptions): void {
  const to = Array.isArray(o.to) ? o.to : o.to.split(',');
  if (to.length === 0 || to.some((n) => n.trim().length === 0))
    throw new MsGineValidationError('to: at least one recipient is required', 'to');
  if (!o.message || o.message.trim().length === 0)
    throw new MsGineValidationError('message is required', 'message');
  if (o.message.length > 1000)
    throw new MsGineValidationError('message must be 1 000 characters or fewer', 'message');
  if (o.from && o.from.length > 11)
    throw new MsGineValidationError('from (sender ID) must be 11 characters or fewer', 'from');
}
